R/status.R

```diff
@@ -155,141 +209,55 @@ renv_status_check_unknown_sources <- function(project, lockfile) {
-
-    # if these packages are in the lockfile, report those records
-    if (all(missing %in% names(lockfile))) {
-
-      records <- keep(lockfile, missing)
-      renv_pretty_print_records(
-        records,
-        preamble  = lockmsg,
-        postamble = restoremsg
-      )
-
-      return(FALSE)
-
-    }
-
-    # otherwise, try to report intelligently
-    postamble <- if (any(missing %in% names(lockfile))) {
-      c(restoremsg, statusmsg)
-    } else {
-      c(installmsg, statusmsg)
-    }
-
-    renv_pretty_print(
-      sort(missing),
-      preamble  = usedmsg,
-      postamble = postamble
-    )
-
-    return(FALSE)
-
-  }
-
-  # flag set to FALSE if any of the below checks report out-of-sync
-  ok <- TRUE
-
-  # not installed/recorded/used ------------------------------------------------
-  records <- lockfile %>%
-    exclude(names(library)) %>%
-    keep(packages)
+renv_status_check_consistent <- function(lockfile, library, used) {
 
-  if (length(records)) {
+  packages <- sort(unique(c(names(library), names(lockfile), used)))
 
-    renv_pretty_print_records(
-      records,
-      "The following package(s) are recorded in the lockfile, but not installed:",
-      "Use `renv::restore()` to install these packages."
-    )
-
-    ok <- FALSE
-
-  }
-
-  # installed/not recorded/used ------------------------------------------------
-  records <- library %>%
-    exclude(names(lockfile)) %>%
-    keep(packages)
-
-  if (length(records)) {
-
-    renv_pretty_print_records(
-      records,
-      "The following package(s) are installed, but not recorded in the lockfile:",
-      "Use `renv::snapshot()` to add these packages to the lockfile."
-    )
-
-    ok <- FALSE
+  status <- data.frame(
+    package = packages,
+    installed = packages %in% names(library),
+    recorded = packages %in% names(lockfile),
+    used = packages %in% used
+  )
 
+  pkg_ok <- status$installed & (status$used == status$recorded)
+  if (all(pkg_ok)) {
+    return(TRUE)
   }
 
-  # */recorded/not used --------------------------------------------------------
-  records <- lockfile %>% exclude(packages)
-  if (length(records)) {
+  if (renv_verbose()) {
+    # If any packages are not installed, we don't know for sure what's used
+    # because our dependency graph is incomplete
+    missing <- any(!status$installed)
 
-    renv_pretty_print_records(
-      records,
-      preamble =
-        "The following packages are recorded in the lockfile, but do not appear to be used in this project:",
-      postamble =
-        "Use `renv::snapshot()` if you'd like to remove these packages from the lockfile."
-    )
-
-    ok <- FALSE
+    issues <- status[!pkg_ok, , drop = FALSE]
+    issues$installed <- ifelse(issues$installed, "y", "n")
+    issues$recorded <- ifelse(issues$recorded, "y", "n")
+    issues$used <- ifelse(issues$used, "y", if (missing) "?" else "n")
 
+    writef("The following package(s) are inconsistent:")
```

Comment:
```suggestion
    writef("The following package(s) are in an inconsistent state:")
```

Reply:
Maybe pedantic but the package itself isn't inconsistent; it's the way it appears to be used / recorded in the project that is?